### PR TITLE
Add the optional S3bucket name

### PR DESCRIPTION
**Pull request by alan-wu**
# Description

Add the optional s3 bucket name, this is a combination of multiple branches and pull request- 
https://github.com/nih-sparc/sparc-api/tree/s3bucket_refactor
https://github.com/nih-sparc/sparc-api/pull/160
More works are on the way for the front end and new pull request will be added subsequently.

## Type of change

Delete those that don't apply.

- [ ] Bug fix (non-breaking change which fixes an issue)

# How Has This Been Tested?

I ran the unit tests and they passed.

# Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added unit tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: app/main.py [hunk 1]
```diff
@@ -295,7 +296,10 @@ def extract_thumbnail_from_xml_file():
     query_args = request.args
     if 'path' not in query_args:
         return abort(400, description=f"Query arguments are not valid.")
+    # if 's3BucketName' not in query_args:
+    #     return abort(400, description=f"Query arguments must include 's3BucketName'.")
 
+    s3BucketName = query_args.get("s3BucketName") or Config.S3_BUCKET_NAME
```
Comment (hsorby) [inline app/main.py:302]: I think it is better to get from a dictionary like this in Python: `query_args.get("s3BucketName", Config.S3_BUCKET_NAME)`

File: app/main.py [hunk 2]
```diff
@@ -295,7 +296,10 @@ def extract_thumbnail_from_xml_file():
     query_args = request.args
     if 'path' not in query_args:
         return abort(400, description=f"Query arguments are not valid.")
+    # if 's3BucketName' not in query_args:
+    #     return abort(400, description=f"Query arguments must include 's3BucketName'.")
```
Comment (hsorby) [inline app/main.py:300]: These comments should be removed.

File: app/main.py [hunk 3]
```diff
@@ -391,24 +402,27 @@ def get_discover_path():
 # important to keep the relative <path> for accessing
 # other required files.
 @app.route("/s3-resource/<path:path>")
-def direct_download_url(path):
+def direct_download_url(path, bucket_name=Config.S3_BUCKET_NAME):
```
Comment (hsorby) [inline app/main.py:405]: Not sure why the bucket_name appears in the parameter list, it is not consistent with any other similar API.

File: app/main.py [hunk 4]
```diff
@@ -391,24 +402,27 @@ def get_discover_path():
...
+
     try:
         head_response = s3.head_object(
-            Bucket=Config.S3_BUCKET_NAME,
+            Bucket=s3BucketName,
             Key=path,
             RequestPayer="requester"
         )
+        content_length = head_response.get('ContentLength', Config.DIRECT_DOWNLOAD_LIMIT)
+        if content_length and content_length > Config.DIRECT_DOWNLOAD_LIMIT:  # 20 MB
```
Comment (hsorby) [inline app/main.py:417]: Surely `content_length` is always defined here? If the `ContentLength` header is not set then you are happy to download a file of any size?

**Review by hsorby — CHANGES_REQUESTED** [review]

File: app/main.py (hunk 3 above)
Comment (alan-wu) [reply app/main.py:405]: It is not consistent as it is called by directly by simulation_ui_file as well thus the parameter.

File: app/main.py (hunk 4 above)
Comment (alan-wu) [reply app/main.py:417]: Hmmm, I moved this block into the try block without looking into the logic but you are right, I will change that.

File: app/main.py (hunk 3 above)
Comment (alan-wu) [reply app/main.py:405]: I will add the parameter for all the other similar api.

**Review by hsorby — COMMENTED** [review]
So now for each API we can either use the query argument `bucket_name` or `s3BucketName`?  Do we need both, is one not sufficient?

**Comment by egauzens** [comment]
> So now for each API we can either use the query argument `bucket_name` or `s3BucketName`? Do we need both, is one not sufficient?

I agree, I think bucket_name can be removed. Also I think Config.S3_BUCKET_NAME should be renamed to DEFAULT_S3_BUCKET_NAME

Either that, or we force the requests to have a s3BucketName passed in as an argument to that we can remove the config var entirely

**Review by egauzens — CHANGES_REQUESTED** [review]
I think bucket_name can be removed and the endpoints should be changed to contain either:

if 's3BucketName' not in query_args:
        return abort(400, description=f"Query arguments must include 's3BucketName'.")
s3BucketName = query_args.get("s3BucketName")

or:

s3BucketName = request.args.get("s3BucketName") or Config.DEFAULT_S3_BUCKET_NAME

**Comment by alan-wu** [comment]
@egauzens @hsorby I am trying to understand the issue here. The argument - bucket_name can be useful, it provides the flexibility to call these methods in main.py with the bucket name set. Take a look at the simulation_ui_file method and it calls once of these methods, we can not set the bucket name without the argument.

**Comment by egauzens** [comment]
> @egauzens @hsorby I am trying to understand the issue here. The argument - bucket_name can be useful, it provides the flexibility to call these methods in main.py with the bucket name set. Take a look at the simulation_ui_file method and it calls once of these methods, we can not set the bucket name without the argument.

Ah ok I did not see it was used by other endpoints in the api. In that case it makes sense

**Comment by alan-wu** [comment]
> So now for each API we can either use the query argument `bucket_name` or `s3BucketName`? Do we need both, is one not sufficient?

@hsorby bucket_name provides the default value and can be used by other methods internally while s3BucketName is the value of the url query or default value when called through the route.

File: tests/test_api.py [hunk 5]
```diff
@@ -22,6 +22,14 @@ def test_direct_download_url_small_file(client):
     assert r.status_code == 200
     assert b"proximal colon" in r.data
 
+def test_direct_download_url_new_bucket_file(client):
+    #whil
```
Comment (hsorby) [inline tests/test_api.py:26]: Remove.

**Review by hsorby — CHANGES_REQUESTED** [review]
Well, certainly things could be refactored to be better. Not convinced with the default arguments, only one `bucket_name` is reused internally.

One tidy up change that needs to be dealt with.

File: tests/test_api.py (hunk 5 above)
Comment (alan-wu) [reply tests/test_api.py]: I understand that, I can revert those ones but I am hoping to keep them consistent.

**Review by hsorby — APPROVED** [review]

**Review by egauzens — APPROVED** [review]

**Review by egauzens — APPROVED** [review]

**Pull request merged**